<a href="https://colab.research.google.com/github/BrianGakungi/Brian-Njagi-Autolib-Analysis/blob/main/John_Njagi_Notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# import pandas and numpy
import pandas as pd
import numpy as np


In [ ]:
# Loading dataset and previewing
df = pd.read_csv('/content/Autolib_dataset.csv')
df.head()

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23

[5 rows x 25 columns]

**CLEANING OUR DATASET :**

**1. VALIDITY**

In [ ]:
# Drop irrelevant columns
df.drop(["Displayed comment", "Scheduled at"], axis=1, inplace=True)

In [ ]:
df.head()

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23

[5 rows x 23 columns]

In [ ]:
# Remove trailing and leading spaces
df['Address'] = df['Address'].str.strip()
df['City'] = df['City'].str.strip()
df['ID'] = df['ID'].str.strip()
df['Public name'] = df['Public name'].str.strip()

**2. COMPLETENESS**


In [ ]:
# check for missing values
df.isnull().sum()

Address                0
Cars                   0
Bluecar counter        0
Utilib counter         0
Utilib 1.4 counter     0
Charge Slots           0
Charging Status        0
City                   0
ID                     0
Kind                   0
Geo point              0
Postal code            0
Public name            0
Rental status          0
Slots                  0
Station type           0
Status                 0
Subscription status    0
year                   0
month                  0
day                    0
hour                   0
minute                 0
dtype: int64

In [ ]:
# drop duplicates
df.drop_duplicates


<bound method DataFrame.drop_duplicates of                                  Address  Cars  ...  hour  minute
0                    2 Avenue de Suffren     0  ...    11      43
1              145 Rue Raymond Losserand     6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy     3  ...    20      14
3                 51 Rue Eugène Oudiné     3  ...     4      37
4     6 avenue de la Porte de Champerret     3  ...    17      23
...                                  ...   ...  ...   ...     ...
4995                 8 avenue Ménelotte     2  ...    11      26
4996                   37 rue de Dantzig     4  ...    16      56
4997                      142 rue du Bac     1  ...     7       1
4998         2 avenue du Val de Fontenay     2  ...    17      27
4999                 5 Rue Henry Brisson     4  ...    10       2

[5000 rows x 23 columns]>

**3. UNIFORMITY**

In [ ]:
# checking datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 23 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Address              5000 non-null   object
 1   Cars                 5000 non-null   int64 
 2   Bluecar counter      5000 non-null   int64 
 3   Utilib counter       5000 non-null   int64 
 4   Utilib 1.4 counter   5000 non-null   int64 
 5   Charge Slots         5000 non-null   int64 
 6   Charging Status      5000 non-null   object
 7   City                 5000 non-null   object
 8   ID                   5000 non-null   object
 9   Kind                 5000 non-null   object
 10  Geo point            5000 non-null   object
 11  Postal code          5000 non-null   int64 
 12  Public name          5000 non-null   object
 13  Rental status        5000 non-null   object
 14  Slots                5000 non-null   int64 
 15  Station type         5000 non-null   object
 16  Status

In [ ]:
# export to csv
df.to_csv('df4.csv')

**CONNECTING TO OUR DATABASE**

In [ ]:
# We will first load the sql extension into our environment
#
%load_ext sql

# Then connect to our in memory sqlite database
# NB: This database will cease to exist as soon as the database connection is closed. 
#
%sql sqlite://

'Connected: @None'

In [ ]:
# Importing the pandas library
# We will use a function read_csv from pandas to read our datasets as shown
#
import pandas as pd 

In [ ]:
# Loading our table from the respective CSV files
# We will export the Households dataset that we have cleaned using python
#
with open('/content/df4.csv','r') as f:
    Electric = pd.read_csv(f, index_col=0, encoding='utf-8')
%sql DROP TABLE if EXISTS Electric;
%sql PERSIST Electric;
%sql SELECT * FROM Electric LIMIT 5;

 * sqlite://
Done.
 * sqlite://
 * sqlite://
Done.


[(0, '2 Avenue de Suffren', 0, 0, 0, 0, 0, 'nonexistent', 'Paris', 'paris-suffren-2', 'STATION', '48.857, 2.2917', 75015, 'Paris/Suffren/2', 'operational', 2, 'station', 'ok', 'nonexistent', 2018, 4, 8, 11, 43),
 (1, '145 Rue Raymond Losserand', 6, 6, 0, 0, 0, 'operational', 'Paris', 'paris-raymondlosserand-145', 'STATION', '48.83126, 2.313088', 75014, 'Paris/Raymond Losserand/145', 'operational', 0, 'station', 'ok', 'nonexistent', 2018, 4, 6, 7, 24),
 (2, '2 Avenue John Fitzgerald Kennedy', 3, 3, 0, 2, 0, 'operational', 'Le Bourget', 'lebourget-johnfitzgeraldkennedy-2', 'STATION', '48.938103, 2.4286035', 93350, 'Le Bourget/John Fitzgerald Kennedy/2', 'operational', 1, 'station', 'ok', 'nonexistent', 2018, 4, 3, 20, 14),
 (3, '51 Rue Eugène Oudiné', 3, 3, 1, 0, 1, 'operational', 'Paris', 'paris-eugeneoudine-51', 'STATION', '48.8250327, 2.3725162', 75013, 'Paris/Eugène Oudiné/51', 'operational', 2, 'station', 'ok', 'nonexistent', 2018, 4, 4, 4, 37),
 (4, '6 avenue de la Porte de Champerret', 3, 3, 0, 0, 0, 'nonexistent', 'Paris', 'paris-portedechamperret-6', 'PARKING', '48.8862632, 2.2874511', 75017, 'Paris/Porte de Champerret/6', 'operational', 3, 'station', 'ok', 'nonexistent', 2018, 4, 8, 17, 23)]

**Research Question**

**Identify the most popular hour of the day for picking up a shared electric car (Bluecar) in the city of Paris over the month of April 2018.** 

In [ ]:
%%sql
SELECT hour, COUNT(hour) from Electric
where "Bluecar counter" = 0 and "City" = "Paris"
GROUP BY hour
ORDER BY COUNT(hour) Desc limit 1;

 * sqlite://
Done.


[(2, 54)]

In [ ]:
# What is the most popular hour for returning cars?
%%sql
SELECT hour, COUNT(hour) from Electric
where "Slots" = 0 
GROUP BY hour
ORDER BY COUNT(hour) DESC LIMIT 1 ;

 * sqlite://
Done.


[(21, 89)]

In [ ]:
# What station is the most popular overall?
%%sql
SELECT "Public name", COUNT("Public name") from Electric
where "Cars" = 0 
GROUP BY "Public name"
ORDER BY COUNT("Public name") DESC limit 1;

 * sqlite://
Done.


[('Paris/Grenelle 31/Parking', 11)]

In [ ]:
# What station is the most popular at the most popular picking hour?
%%sql
SELECT "Public name", COUNT("Public name") from Electric
where "Cars" = 0 and hour = 2
GROUP BY "Public name" 
ORDER BY Count("Public name") DESC;

 * sqlite://
Done.


[('Bièvres/Gare/12', 1),
 ('Bourg-la-Reine/André Theuriet/1', 1),
 ('Champigny-sur-Marne/Pierre Renaudel/2', 1),
 ('Gennevilliers/De Gaulle/23', 1),
 ('Joinville-le-Pont/Mozart/18', 1),
 ('Limeil-Brévannes/Lavoisier/12', 1),
 ('Pantin/Pré Saint Gervais/1', 1),
 ('Paris/Artois/11', 1),
 ('Paris/Bac/142', 1),
 ('Paris/Batignolles/9', 1),
 ('Paris/Bernardins/23', 1),
 ('Paris/Bosquet/41', 1),
 ('Paris/Boucry/35', 1),
 ('Paris/Château Landon/24', 1),
 ('Paris/Claude Vellefaux/3', 1),
 ('Paris/Commandant Lamy/2', 1),
 ('Paris/Courcelles/210', 1),
 ('Paris/Dante/1', 1),
 ('Paris/Dutuit/1', 1),
 ('Paris/Etoile Foch/Parking', 1),
 ('Paris/Fleurs/23', 1),
 ('Paris/Four/38', 1),
 ('Paris/Francois 1er/7', 1),
 ('Paris/François 1er/38', 1),
 ('Paris/Friedland/42', 1),
 ('Paris/Garibaldi Cambronne/Parking', 1),
 ('Paris/Georges Mandel/42', 1),
 ('Paris/Grande Armée/81', 1),
 ('Paris/Halles/3', 1),
 ('Paris/Haussmann/6', 1),
 ('Paris/Henri IV/46', 1),
 ('Paris/Invalides/21', 1),
 ('Paris/Legendre/22', 1),
 ('Paris/Linné/18', 1),
 ('Paris/Louvre/31', 1),
 ('Paris/Madame/18', 1),
 ('Paris/Madeleine/8', 1),
 ('Paris/Megisserie/20', 1),
 ('Paris/Michelet/6', 1),
 ('Paris/Milton/12', 1),
 ('Paris/Montparnasse Raspail/Parking', 1),
 ("Paris/Nonnains d'Hyères/1", 1),
 ('Paris/Ordener/1', 1),
 ('Paris/Paix/24', 1),
 ('Paris/Pierre Chausson/1', 1),
 ('Paris/Quatre septembre/10', 1),
 ('Paris/Raspail/136', 1),
 ('Paris/Rennes Montparnasse/Parking', 1),
 ('Paris/Rennes/105', 1),
 ('Paris/Rocroy/15', 1),
 ('Paris/République/17', 1),
 ('Paris/Saint-Antoine/199', 1),
 ('Paris/Saint-Germain des Prés/Parking', 1),
 ('Paris/Saint-Jacques/221', 1),
 ('Paris/Saint-Maur/36', 1),
 ('Paris/Ternes/73', 1),
 ('Paris/Trudaine/26', 1),
 ('Paris/Uzès/1', 1),
 ('Paris/Vincent Auriol/81', 1),
 ('Paris/Vivienne/43', 1),
 ('Paris/Voltaire/8', 1),
 ('Rueil-Malmaison/Lavoisier/3', 1),
 ('Saint-Ouen/Frida Kahlo/20', 1),
 ('Versailles/Bonne Aventure/68', 1)]

In [ ]:
# What postal code is the most popular for picking up Blue cars overall?
%%sql
SELECT "Postal code", COUNT("Postal code") from Electric
where "Bluecar counter" = 0 
GROUP BY "Postal code"
ORDER BY COUNT("Postal code") Desc limit 1;

 * sqlite://
Done.


[(75008, 94)]

In [ ]:
# Does the most popular station belong to that postal code?
%%sql
Select "Postal code" from Electric
where "Public name" = "Paris/Grenelle 31/Parking";

 * sqlite://
Done.


[(75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,)]

No the most popular station does not belong to that postal code since the postal code for the most popular station is 75015 and the postal code for the station that's most popluar for picking blue cars is 75008

In [ ]:
# What postal code is the most popular for picking up Blue cars at the most popular picking hour?
%%sql
SELECT "Postal code", COUNT("Postal code") from Electric
where "Bluecar counter" = 0 and hour = 2
GROUP BY "Postal code"
ORDER BY COUNT("Postal code") Desc limit 1;

 * sqlite://
Done.


[(75006, 7)]

In [ ]:
# Does the most popular station belong to that postal code?
%%sql
Select "Postal code" from Electric
where "Public name" = "Paris/Grenelle 31/Parking";

 * sqlite://
Done.


[(75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,)]

No the most popular station does not belong to that postal code since the postal code for the most popular station is 75015 and the postal code for the station that's most popluar for picking blue cars at the most popular picking hour is 75006

Do the results change if you consider Utilib and Utilib 1.4 instead of Blue cars? 

In [ ]:
# What postal code is the most popular for picking up Utilib cars overall?
%%sql
SELECT "Postal code", COUNT("Postal code") from Electric
where "Utilib counter" = 0 
GROUP BY "Postal code"
ORDER BY COUNT("Postal code") Desc limit 1;

 * sqlite://
Done.


[(75015, 257)]

In [ ]:
# Does the most popular station belong to that postal code?
%%sql
Select "Postal code" from Electric
where "Public name" = "Paris/Grenelle 31/Parking";

 * sqlite://
Done.


[(75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,)]

Yes the most popular station belongs to that postal code

In [ ]:
# What postal code is the most popular for picking up Utilib 1.4 cars overall?
%%sql
SELECT "Postal code", COUNT("Postal code") from Electric
where "Utilib 1.4 counter" = 0 
GROUP BY "Postal code"
ORDER BY COUNT("Postal code") Desc limit 1;

 * sqlite://
Done.


[(75015, 245)]

In [ ]:
# Does the most popular station belong to that postal code?
%%sql
Select "Postal code" from Electric
where "Public name" = "Paris/Grenelle 31/Parking";

 * sqlite://
Done.


[(75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,)]

Yes the most popular station belongs to that postal code

In [ ]:
# What postal code is the most popular for picking up Utilib cars at the most popular picking hour?
%%sql
SELECT "Postal code", COUNT("Postal code") from Electric
where "Utilib counter" = 0 and hour = 2
GROUP BY "Postal code"
ORDER BY COUNT("Postal code") Desc limit 1;

 * sqlite://
Done.


[(75005, 10)]

In [ ]:
# Does the most popular station belong to that postal code?
%%sql
Select "Postal code" from Electric
where "Public name" = "Paris/Grenelle 31/Parking";

 * sqlite://
Done.


[(75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,)]

No the most popular station does not belong to that postal code since the postal code for the most popular station is 75015 and the postal code for the station that's most popluar for picking Utilib cars at the most popular picking hour is 75005

In [ ]:
# What postal code is the most popular for picking up Utilib 1.4 cars at the most popular picking hour?
%%sql
SELECT "Postal code", COUNT("Postal code") from Electric
where "Utilib 1.4 counter" = 0 and hour = 2
GROUP BY "Postal code"
ORDER BY COUNT("Postal code") Desc limit 1;

 * sqlite://
Done.


[(75011, 11)]

In [ ]:
# Does the most popular station belong to that postal code?
%%sql
Select "Postal code" from Electric
where "Public name" = "Paris/Grenelle 31/Parking";

 * sqlite://
Done.


[(75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,),
 (75015,)]

No the most popular station does not belong to that postal code since the postal code for the most popular station is 75015 and the postal code for the station that's most popluar for picking Utilib 1.4 cars at the most popular picking hour is 75011

**YES THE RESULTS CHANGE IF YOU CONSIDER UTILIB AND UTILIB 1.4 CARS**

Here's the link to the Github repository
https://github.com/BrianGakungi/John-Njagi-Autolib-Analysis.git


Here's the link to the JIRA Board 
https://johngakungi.atlassian.net/jira/software/projects/M1/boards/1